### ISSUE TITLE
@Timed annotation on resource super class method throws exception on init

### ISSUE THREAD

**Issue by jkytomaki** [dropwizard/metrics#746]
If multiple jersey resource classes extend a common base class, and that base class has one resource method with @Timed annotation, Dropwizard's initialization will fail with e.g.

```
java.lang.IllegalArgumentException: duplicate key: public javax.ws.rs.core.Response 
com.codahale.metrics.jersey2.
InstrumentedResourceMethodApplicationListenerTest$ResourceBase.getCommonThings()
```

The same resource classes work with Metrics 2.2.0.

The problem seems to be in the way the map key is selected for registered metrics in InstrumentedResourceMethodApplicationListener's registerTimedAnnotations method (and registerMeteredAnnotations as well). A java.lang.reflect.Method is used as a key, but for the resource method inherited from the base class, it is the same instance for all of the subclasses and thus the initialization will fail on key duplicity if there are more than one subclass.

Here is an example junit test that should replicate the problem (the test method is empty, but the test should still fail on initialization).

``` java
public class InstrumentedResourceMethodApplicationListenerTest extends JerseyTest{

  @Override
  protected Application configure() {
    MetricRegistry registry = new MetricRegistry();
    InstrumentedResourceMethodApplicationListener listener = new
            InstrumentedResourceMethodApplicationListener(registry);

    ResourceConfig rc = new ResourceConfig(FooResource.class, BarResource.class);
    rc.register(listener);
    return rc;
  }

  public abstract class ResourceBase {
    @GET
    @Path("/common-things")
    @Timed
    public Response getCommonThings(){
      return Response.ok().build();
    }
  }

  @Path("/foos")
  public class FooResource extends ResourceBase {
    @GET
    public Response getFoos(){
      return Response.ok().build();
    }
  }

  @Path("/bars")
  public class BarResource extends ResourceBase {
    @GET
    public Response getBars(){
      return Response.ok().build();
    }
  }

  @Test
  public void test() throws InterruptedException {
  }

}
```

**Comment by marshallpierce** [comment]
Could you give my alternate jersey2 integration a try? It should not suffer from this problem. https://bitbucket.org/marshallpierce/jersey2-metrics

**Comment by OlaviSau** [comment]
By the way check out HttpContext for automatic paths

**Comment by mtakaki** [comment]
This is actually an issue on the metrics package and it's being tracked under: https://github.com/dropwizard/metrics/issues/859

I submitted a pull request suggesting a fix for this issue, keeping the same behavior (using the class name and method) but using the child class name instead of the parent: https://github.com/dropwizard/metrics/pull/861

**Comment by tostrowski** [comment]
:+1: since this bit us as well.

**Comment by mtakaki** [comment]
@tostrowski, feel free to copy my fix while it's not accepted (or rejected) on the pull request: https://github.com/dropwizard/metrics/pull/861

It fixes the issue and works as you would expect: each child resource gets its own entry in the timers group. It also fixes the issue for `@ExceptionMetered`, that will be under the meters group.

To add it to your project, just copy the file `com.codahale.metrics.jersey2.InstrumentedResourceMethodApplicationListener` under your project. The class loader will put your classes before the metrics one and it will work. That's what I'm doing right now, as I can't afford to wait for this to be pulled.

**Comment by rferreira** [comment]
+1 on this bug

**Comment by gnmerritt** [comment]
Thanks @mtakaki - this fixes the issue for Dropwizard 0.9.1 as well

**Comment by mtakaki** [comment]
@gnmerritt I'm glad that it helped you!
